File: examples/threads/InterruptTest.java

/**
 * Shows how to interrupt a sleeping thread. Why does this program never end
 * even though the main method terminates?
 * 
 * @author amit
 *
 */
public class InterruptTest implements Runnable {

	public static void main(String[] args) throws Exception {
		Thread sleepyThread = new Thread(new InterruptTest());
		sleepyThread.setName("SleepyThread");
		sleepyThread.start();
		// now we two threads running, the main thread and the sleepy thread,
		// which goes to sleep after printing a message.

		Thread.sleep(3000);// put main thread to sleep for a while
		sleepyThread.interrupt();// interrupt sleepyThread's beauty sleep

		Thread.sleep(500);// put main thread to sleep for a while
		sleepyThread.interrupt();// interrupt sleepyThread's beauty sleep
	}

	public void run() {
		Thread me = Thread.currentThread();
		while (true) {
			try {
				System.out.println(me.getName() + ": zzzz...");
				Thread.sleep(5 * 1000); // in millisecs
			} catch (InterruptedException e) {
				System.out.println(me.getName() + ": argh! let me sleep #$@!");
			}
		}
	}
}

File: examples/other-languages/kotlin/threads/InterruptTest.kt
// We can omit the string array if it's not used.
fun main() {
    // Example use case of Kotlin's special syntax. The Runnable interface has
    // a single method (void run()), so we can replace an instance of a class
    // with a lambda of type () -> Unit. Kotlin also allows us to omit the
    // parentheses when calling a function whose last argument is a lambda.
    val sleepyThread = Thread {
        while (true) {
            val name = Thread.currentThread().name
            try {
                println("$name: zzzz...")
                Thread.sleep(5000)
            } catch (e: InterruptedException) {
                println("$name: argh! let me sleep #\$@!")
            }
        }
    }
    sleepyThread.name = "SleepyThread"
    sleepyThread.start()
    // Now we have two threads, the main thread and the sleepy thread

    val name = Thread.currentThread().name
    Thread.sleep(5000)
    println("$name: disturbing the peacefully sleeping thread")
    sleepyThread.interrupt()

    Thread.sleep(5000)
    println("$name: wake up!")
    sleepyThread.interrupt()
}